creusot/src/translation/function/terminator.rs

```diff
@@ -71,16 +75,104 @@ impl<'tcx> BodyTranslator<'_, 'tcx> {
+                                )
+                                .emit();
+                        }
+                    } else if self
+                        .tcx
+                        .is_diagnostic_item(Symbol::intern("ghost_box_new"), fun_def_id)
+                    {
+                        self.ctx
+                            .error(*fn_span, "cannot create a ghost variable in program context")
+                            .with_span_suggestion(
+                                *fn_span,
+                                "try wrapping this expression in `gh!` instead",
+                                format!(
+                                    "gh!{{ {} }}",
+                                    self.ctx
+                                        .sess
+                                        .source_map()
+                                        .span_to_snippet(args[0].span)
+                                        .unwrap()
+                                ),
+                                rustc_errors::Applicability::MachineApplicable,
+                            )
+                            .emit();
+                    } else {
+                        // Check and reject instantiation of a <T: Deref> with a ghost parameter.
+                        let deref_trait_id =
+                            self.tcx.get_diagnostic_item(Symbol::intern("Deref")).unwrap();
+                        let infer_ctx = self.tcx.infer_ctxt().build();
+                        for bound in func_param_env.caller_bounds() {
+                            let Some(trait_clause) = bound.as_trait_clause() else { continue };
+                            if trait_clause.def_id() != deref_trait_id {
+                                continue;
+                            }
+                            let ty = trait_clause.self_ty().skip_binder();
+                            let caller_ty = ty::EarlyBinder::bind(trait_clause.self_ty())
+                                .instantiate(self.tcx, subst)
+                                .skip_binder();
+                            let deref_in_callee = infer_ctx
+                                .type_implements_trait(
+                                    deref_trait_id,
+                                    std::iter::once(ty),
+                                    func_param_env,
+                                )
+                                .may_apply();
+                            let ghost_in_caller = self.is_ghost_box(caller_ty);
+                            let ghost_in_callee = self.is_ghost_box(ty);
+                            if deref_in_callee && ghost_in_caller && !ghost_in_callee {
+                                self.ctx.error(*fn_span, &format!("Cannot instantiate a generic type {ty} implementing `Deref` with the ghost type {caller_ty}")).emit();
+                            }
+                        }
+                    }
+                }
```

Comment:
Can we refactor  this into  helper function(s)?

Reply:
Done